File: tests/core-commands/CommandConformance.swift
import Foundation

func encoded(_ object: Any) throws -> String {
    let data = try JSONSerialization.data(withJSONObject: object, options: [.sortedKeys, .withoutEscapingSlashes])
    return String(decoding: data, as: UTF8.self)
}

@main
struct CommandConformance {
    static func main() throws {
        let args = CommandLine.arguments
        guard args.count == 4 else { fatalError("usage: command_conformance INPUT COMMANDS OUTPUT") }
        let source = try String(contentsOfFile: args[1], encoding: .utf8)
        let commands = try JSONSerialization.jsonObject(with: Data(contentsOf: URL(fileURLWithPath: args[2])))
        let session = try NativeSession.open(source: source)
        let opened = try session.exportJson()
        let begin = try session.beginTransactionJson(request: encoded(["version": 1, "expectedRevision": 0]))
        let update = try session.updateTransactionJson(request: encoded(["version": 1, "transactionId": 1, "commands": commands]))
        let draft = try session.exportJson()
        let commit = try session.commitTransactionJson(request: encoded(["version": 1, "transactionId": 1]))
        let committed = try session.exportJson()
        let didUndo = try session.undo()
        precondition(didUndo)
        let undone = try session.exportJson()
        let didRedo = try session.redo()
        precondition(didRedo)
        let redone = try session.exportJson()
        let reopened = try NativeSession.open(source: session.exportDurableJson()).exportJson()
        let stale = try session.cancelTransactionJson(request: encoded(["version": 1, "transactionId": 1]))
        let commandList = commands as! [[String: Any]]
        let first = commandList[0]["id"] as! String
        let package = try JSONSerialization.jsonObject(with: Data(source.utf8)) as! [String: Any]
        let document = package["document"] as! [String: Any]
        let global = document["global"] as! [String: Any]
        let seed = global["seed"] as! Int
        let cancelBegin = try session.beginTransactionJson(request: encoded(["version": 1, "expectedRevision": try session.revision()]))
        let cancelUpdate = try session.updateTransactionJson(request: encoded(["version": 1, "transactionId": 2,
            "commands": [["type": "patch_global", "patch": ["seed": seed + 2]]]]))
        let cancelDraft = try session.exportJson()
        let failed = try session.updateTransactionJson(request: encoded(["version": 1, "transactionId": 2,
            "commands": [
                ["type": "bridge", "capability": "web:layer-property",
                 "target": ["scope": "layer_field", "id": first, "field": "cancelledNull"], "value": NSNull()],
                ["type": "patch_global", "patch": ["aspect": "invalid"]],
            ]]))
        let failedDraft = try session.exportJson()
        let cancel = try session.cancelTransactionJson(request: encoded(["version": 1, "transactionId": 2]))
        let cancelled = try session.exportJson()
        let legacyCommand = try encoded(["type": "edit_layer", "id": first, "patch": ["name": "P03 legacy"]])
        let didLegacyEdit = try session.execute(commandJson: legacyCommand)
        precondition(didLegacyEdit)
        let legacy = try session.exportJson()
        let mixedBegin = try session.beginTransactionJson(request: encoded(["version": 1, "expectedRevision": try session.revision()]))
        let mixedUpdate = try session.updateTransactionJson(request: encoded(["version": 1, "transactionId": 3,
            "commands": [["type": "patch_global", "patch": ["seed": seed + 3]]]]))
        let mixedCommit = try session.commitTransactionJson(request: encoded(["version": 1, "transactionId": 3]))
        let mixedAfter = try session.exportJson()
        let mixedDidUndoTransaction = try session.undo()
        precondition(mixedDidUndoTransaction)
        let mixedUndoTransaction = try session.exportJson()
        let mixedDidUndoLegacy = try session.undo()
        precondition(mixedDidUndoLegacy)
        let mixedUndoLegacy = try session.exportJson()
        let mixedDidRedoLegacy = try session.redo()
        precondition(mixedDidRedoLegacy)
        let mixedRedoLegacy = try session.exportJson()
        let mixedDidRedoTransaction = try session.redo()
        precondition(mixedDidRedoTransaction)
        let mixedRedoTransaction = try session.exportJson()
        let output: [String: String] = ["opened": opened, "begin": begin, "update": update, "draft": draft,
                                         "commit": commit, "committed": committed, "undone": undone,
                                         "redone": redone, "reopened": reopened, "stale": stale,
                                         "cancelBegin": cancelBegin, "cancelUpdate": cancelUpdate, "cancelDraft": cancelDraft,
                                         "failed": failed, "failedDraft": failedDraft, "cancel": cancel, "cancelled": cancelled,
                                         "legacy": legacy, "mixedBegin": mixedBegin, "mixedUpdate": mixedUpdate,
                                         "mixedCommit": mixedCommit, "mixedAfter": mixedAfter,
                                         "mixedUndoTransaction": mixedUndoTransaction, "mixedUndoLegacy": mixedUndoLegacy,
                                         "mixedRedoLegacy": mixedRedoLegacy, "mixedRedoTransaction": mixedRedoTransaction]
        try JSONSerialization.data(withJSONObject: output, options: [.sortedKeys, .withoutEscapingSlashes])
            .write(to: URL(fileURLWithPath: args[3]), options: .atomic)
    }
}

File: crates/artifact-core/examples/command_conformance.rs
use artifact_core::DocumentSession;
use serde_json::{Value, json};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 4 {
        return Err("usage: command_conformance INPUT COMMANDS OUTPUT".into());
    }
    let source = std::fs::read_to_string(&args[1])?;
    let commands: Value = serde_json::from_str(&std::fs::read_to_string(&args[2])?)?;
    let mut session = DocumentSession::open(&source)?;
    let opened = session.export_json();
    let begin =
        session.begin_transaction_json(&json!({"version":1,"expectedRevision":0}).to_string());
    let original_aspect = serde_json::from_str::<Value>(&source)?["document"]["global"]["aspect"]
        .as_str()
        .ok_or("fixture needs aspect")?
        .to_owned();
    let alternate_aspect = if original_aspect == "1:1" {
        "4:5"
    } else {
        "1:1"
    };
    let noop_update = session.update_transaction_json(
        &json!({"version":1,"transactionId":1,"commands":[
            {"type":"patch_global","patch":{"aspect":alternate_aspect}},
            {"type":"patch_global","patch":{"aspect":original_aspect}}
        ]})
        .to_string(),
    );
    let update = session.update_transaction_json(
        &json!({"version":1,"transactionId":1,"commands":commands}).to_string(),
    );
    let draft = session.export_json();
    let commit =
        session.commit_transaction_json(&json!({"version":1,"transactionId":1}).to_string());
    let committed = session.export_json();
    assert!(session.undo());
    let undone = session.export_json();
    assert!(session.redo());
    let redone = session.export_json();
    let graph_plan = session.graph_plan_json("__export__")?;
    let reopened = DocumentSession::open(&session.export_durable_json()?)?.export_json();
    let stale =
        session.cancel_transaction_json(&json!({"version":1,"transactionId":1}).to_string());
    let first = commands[0]["id"]
        .as_str()
        .ok_or("first command needs layer id")?;
    let seed = serde_json::from_str::<Value>(&source)?["document"]["global"]["seed"]
        .as_u64()
        .ok_or("fixture needs seed")?;
    let cancel_begin = session.begin_transaction_json(
        &json!({"version":1,"expectedRevision":session.revision()}).to_string(),
    );
    let cancel_update = session.update_transaction_json(
        &json!({"version":1,"transactionId":2,"commands":[
            {"type":"patch_global","patch":{"seed":seed+2}}
        ]})
        .to_string(),
    );
    let cancel_draft = session.export_json();
    let failed = session.update_transaction_json(&json!({"version":1,"transactionId":2,"commands":[
        {"type":"bridge","capability":"web:layer-property","target":{"scope":"layer_field","id":first,"field":"cancelledNull"},"value":null},
        {"type":"patch_global","patch":{"aspect":"invalid"}}
    ]}).to_string());
    let failed_draft = session.export_json();
    let cancel =
        session.cancel_transaction_json(&json!({"version":1,"transactionId":2}).to_string());
    let cancelled = session.export_json();
    assert!(session.execute(
        &json!({"type":"edit_layer","id":first,"patch":{"name":"P03 legacy"}}).to_string()
    )?);
    let legacy = session.export_json();
    let mixed_begin = session.begin_transaction_json(
        &json!({"version":1,"expectedRevision":session.revision()}).to_string(),
    );
    let mixed_update = session.update_transaction_json(
        &json!({"version":1,"transactionId":3,"commands":[
            {"type":"patch_global","patch":{"seed":seed+3}}
        ]})
        .to_string(),
    );
    let mixed_commit =
        session.commit_transaction_json(&json!({"version":1,"transactionId":3}).to_string());
    let mixed_after = session.export_json();
    assert!(session.undo());
    let mixed_undo_transaction = session.export_json();
    assert!(session.undo());
    let mixed_undo_legacy = session.export_json();
    assert!(session.redo());
    let mixed_redo_legacy = session.export_json();
    assert!(session.redo());
    let mixed_redo_transaction = session.export_json();
    let mut structure_update = String::new();
    let mut structure_commit = String::new();
    let mut structure_undo = String::new();
    let mut structure_redo = String::new();
    if serde_json::from_str::<Value>(&source)?["document"]["coldStructureConformance"] == true {
        let current: Value = serde_json::from_str(&session.export_json())?;
        let mut layers = current["document"]["layers"].as_array().unwrap().clone();
        layers.push(json!({"id":"p03-model","kind":"model","unknown":{"nested":null}}));
        let bridge = json!({"type":"bridge_structure","capability":"web:structure","layers":layers,
            "graph":{"present":true,"value":{"edges":[{"id":"p03-edge","fromId":"p03-model","toId":"__export__"}],
            "positions":{},"mergeNodes":[],"colorNodes":[],"unknown":{"nested":null}}}});
        session.begin_transaction_json(
            &json!({"version":1,"expectedRevision":session.revision()}).to_string(),
        );
        structure_update = session.update_transaction_json(
            &json!({"version":1,"transactionId":4,"commands":[bridge]}).to_string(),
        );
        structure_commit =
            session.commit_transaction_json(&json!({"version":1,"transactionId":4}).to_string());
        assert!(session.undo());
        structure_undo = session.export_json();
        assert!(session.redo());
        structure_redo = session.export_json();
    }
    let output = json!({"opened":opened,"begin":begin,"noopUpdate":noop_update,"update":update,"draft":draft,"commit":commit,
        "committed":committed,"undone":undone,"redone":redone,"reopened":reopened,"stale":stale,
        "cancelBegin":cancel_begin,"cancelUpdate":cancel_update,"cancelDraft":cancel_draft,"failed":failed,"failedDraft":failed_draft,
        "cancel":cancel,"cancelled":cancelled,"legacy":legacy,"mixedBegin":mixed_begin,"mixedUpdate":mixed_update,
        "mixedCommit":mixed_commit,"mixedAfter":mixed_after,"mixedUndoTransaction":mixed_undo_transaction,
        "mixedUndoLegacy":mixed_undo_legacy,"mixedRedoLegacy":mixed_redo_legacy,"mixedRedoTransaction":mixed_redo_transaction,
        "structureUpdate":structure_update,"structureCommit":structure_commit,"structureUndo":structure_undo,"structureRedo":structure_redo,
        "graphPlan":graph_plan});
    std::fs::write(&args[3], output.to_string())?;
    Ok(())
}
